File: HLRS/MPI/course/C/totalview/pingpong.c
/*   Program for totalview exercise
     by Thomas Boenisch, HLRS
*/


#include <mpi.h>

main(int argc,char *argv[])
{

   int array[1000][1000];

   char message[55] = "hello, this is the MPI-Course at the HLRS, April 2000";
         
   char message_buf[60];

   int rank, i;

   MPI_Status status;
   
   int ibuf;

   
   MPI_Init(&argc,&argv);

   MPI_Comm_rank(MPI_COMM_WORLD, &rank);

   
   if (rank == 0){
     
     for(i=0;i<54;i++){
       
       MPI_Send(message,i,MPI_CHAR,1,4,MPI_COMM_WORLD);
	 
       MPI_Recv(&ibuf,0,MPI_INT,1,5,MPI_COMM_WORLD,&status);

     }

     MPI_Send(array,1000000,MPI_INT,1,6,MPI_COMM_WORLD);

     MPI_Recv(array,1000000,MPI_INT,1,7,MPI_COMM_WORLD,&status);

   }
   else{

     for(i=0;i<60;i++) message_buf[i]='\0';

     for(i=0;i<54;i++){
	
       MPI_Recv(message_buf,i,MPI_CHAR,0,5,MPI_COMM_WORLD,&status);

       MPI_Send(&ibuf,0,MPI_INT,0,4,MPI_COMM_WORLD);

       printf("%s\n", message_buf);

     }

     MPI_Recv(&ibuf,1000000,MPI_INT,0,6,MPI_COMM_WORLD,&status);
       
     MPI_Send(&ibuf,1000000,MPI_INT,0,7,MPI_COMM_WORLD);

   }
   
   MPI_Finalize();

 }

File: HLRS/MPI/course/F/totalview/pingpong.f90
!   Program for totalview exercise
!   by Thomas Boenisch, HLRS
!

program pingpong

   include 'mpif.h'


   integer array(1000,1000)
   
   character*55 message

   integer rank, ierror

   integer status(MPI_STATUS_SIZE)
   
   integer i_buf(100000)

   
   call mpi_init(ierror)

   call mpi_comm_rank(MPI_COMM_WORLD, rank, ierror)

   
   if (rank == 0) then

      message = 'hello, this is the MPI-Course at the HLRS, April 2000 '
         
      do i = 1, 53 

		 call mpi_send(message,i,MPI_CHAR,1,4,MPI_COMM_WORLD,ierror)

		 call mpi_recv(ibuf,0,MPI_INTEGER,1,5,MPI_COMM_WORLD,status,ierror)

      end do

      call mpi_send(array,1000000,MPI_INTEGER,1,6,MPI_COMM_WORLD,ierror)

      call mpi_recv(array,1000000,MPI_INTEGER,1,7,MPI_COMM_WORLD,status,ierror)


   else

	  do i = 1, 53 

         message = '     '
		 
		 call mpi_recv(message,i,MPI_CHAR,0,5,MPI_COMM_WORLD,status,ierror)

		 call mpi_send(ibuf,0,MPI_INTEGER,0,4,MPI_COMM_WORLD,ierror)

		 write(6,*) message

      end do

      call mpi_recv(ibuf,1000000,MPI_INTEGER,0,6,MPI_COMM_WORLD,status,ierror)

      call mpi_send(ibuf,1000000,MPI_INTEGER,0,7,MPI_COMM_WORLD,ierror)

   endif
   
   call mpi_finalize(ierror)

end
